### PR TITLE
Add go_router_builder

### LINKED ISSUE #66: Generated extension methods for `goNamed` and `pushNamed`
This is an exploratory feature request to add a `build_runner` layer for generating type-safe wrappers around `goNamed` and `pushNamed`.

I'm imagining something like this:

```dart
@Generate(int fid)
GoRoute(
  name: 'family',
  path: '/family/:fid',
  pageBuilder: (context, state) {...},
)
```

Which would create something like this:

```dart
extension on GoRouter {
  void pushFamily(int fid) =>
    push(_lookupNamedRoute('family', {'fid': fid}));

  void goFamily(int fid) =>
    go(_lookupNamedRoute('family', {'fid': fid}));
}
```

Thoughts?

### PR DESCRIPTION
**Pull request by kevmoo**
Fixes https://github.com/csells/go_router/issues/66

Adds two new classes to pkg:go_router

### CODE REVIEW

**Comment by kevmoo** [comment]
- [ ] fix docs related to optional params
- [ ] Add error about using nullable required args - must not be nullable!
- [ ] top-level route thing
- [ ] query param names (keys) as kebab-case
- [ ] add back `pageBuilder` – need to expose random things
- [ ] `expando` idea – ON HOLD! We get a different instance of the state object. This won't work
- [ ] Move example under `go_router_builder`

File: go_router/lib/src/route_data.dart [hunk 1]
```diff
@@ -0,0 +1,120 @@
...
+  ///
+  /// Corresponds to [GoRoute.redirect].
+  String? redirect() => null;
+
+  /// A helper function used by generated code. Should not be used directly.
+  static String $location(String path, {Map<String, String>? queryParams}) =>
+      Uri.parse(path)
+          .replace(
+            queryParameters:
+                // Avoid `?` in generated location if `queryParams` is empty
```
Comment (kevmoo) [inline go_router/lib/src/route_data.dart:45]: @csells – I thought this was fixed. But it's not. FYI

File: go_router/lib/src/route_data.dart [hunk 2]
```diff
@@ -0,0 +1,120 @@
...
+    List<GoRoute> routes = const [],
+  }) {
+    T factoryImpl(GoRouterState state) {
+      final extra = state.extra;
+
+      if (extra is T) {
+        return extra;
+      }
+
+      return (_stateObjectExpando[state] ??= factory(state)) as T;
```
Comment (kevmoo) [inline go_router/lib/src/route_data.dart:63]: It seems that `extra` isn't always populated in some of these flows. Maybe around redirects? Might be worth investigating @csells

Comment (csells) [reply go_router/lib/src/route_data.dart:63]: Can you provide me a minimal repro?

Comment (kevmoo) [reply go_router/lib/src/route_data.dart:63]: The minimal repo is to put in print debugging and see the code paths that come through here where there is no extra.

Guessing it's when redirects are calculated.

Comment (csells) [reply go_router/lib/src/route_data.dart:63]: I'm asking for a set of routes and a set of user actions that produces a null extra parameter.

Comment (kevmoo) [reply go_router/lib/src/route_data.dart:63]: Put a breakpoint on line 63 and run the example in debug mode...that'll do it!

**Comment by chunhtai** [comment]
Since this package has been moved to https://github.com/flutter/packages/tree/main/packages/go_router, can you create a pull request against the flutter/packages instead?

**Pull request closed without merging**